internal/metrics/metrics.go

```diff
@@ -0,0 +1,209 @@
+	ActivePipelines  prometheus.Gauge
+	Up               prometheus.Gauge
+	InvariantRuns    *prometheus.CounterVec
+	AlarmsGenerated  *prometheus.CounterVec
+	NodeErrors       *prometheus.CounterVec
+
+	registry *prometheus.Registry
+	factory  Factory
+	server   *http.Server
+}
+
+var _ Metricer = (*Metrics)(nil)
+
+var Stats Metricer = new(noopMetricer)
+
+var stats = Stats
+
+type StatsKey = string
+type statsKeyType int
+
+const statsKey statsKeyType = iota
+
+// WithContext returns a Metricer from the given context. If no Metricer is found,
+// the default noopMetricer is returned.
+func WithContext(ctx context.Context) Metricer {
+	if ctx == nil {
+		return stats
+	}
+
+	if ctxStats, ok := ctx.Value(statsKey).(Metricer); ok {
+		return ctxStats
+	}
+
+	return stats
+}
+
+// New ... Creates a new metrics server registered with defined custom metrics
+func New(ctx context.Context, cfg *Config) (Metricer, func(), error) {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	registry.MustRegister(collectors.NewGoCollector())
+	factory := With(registry)
+
+	stats = &Metrics{
+		Up: factory.NewGauge(prometheus.GaugeOpts{
+			Namespace: metricsNamespace,
+			Name:      "up",
+			Help:      "1 if the service is up",
+		}),
+		ActiveInvariants: factory.NewGauge(prometheus.GaugeOpts{
+			Name:      "active_invariants",
+			Help:      "Number of active invariants",
+			Namespace: metricsNamespace,
+			Subsystem: SubsystemInvariants,
+		}),
+
+		ActivePipelines: factory.NewGauge(prometheus.GaugeOpts{
+			Name:      "active_pipelines",
+			Help:      "Number of active pipelines",
+			Namespace: metricsNamespace,
+			Subsystem: SubsystemEtl,
+		}),
+
+		InvariantRuns: factory.NewCounterVec(prometheus.CounterOpts{
+			Name:      "invariant_runs_total",
+			Help:      "Number of times a specific invariant has been run",
+			Namespace: metricsNamespace,
+			Subsystem: SubsystemInvariants,
+		}, []string{"invariant"}),
+
+		AlarmsGenerated: factory.NewCounterVec(prometheus.CounterOpts{
+			Name:      "alarms_generated_total",
+			Help:      "Number of total alarms generated for a given invariant",
+			Namespace: metricsNamespace,
+		}, []string{"invariant"}),
+
+		NodeErrors: factory.NewCounterVec(prometheus.CounterOpts{
+			Name:      "node_errors_total",
+			Help:      "Number of node errors caught",
+			Namespace: metricsNamespace,
+		}, []string{"node"}),
+
+		registry: registry,
+		factory:  factory,
+		server:   initServer(cfg, registry),
+	}
+
+	stop := func() {
+		logging.WithContext(ctx).Info("starting to shutdown metrics server")
+		ctx, cancel := context.WithTimeout(context.Background(), serverShutdownTimeout)
+		if err := (stats).(*Metrics).Shutdown(ctx); err != nil {
+			logging.WithContext(ctx).Error("failed to shutdown metrics server: %v", zap.Error(err))
+			panic(err)
+		}
+		defer cancel()
+	}
+
+	return stats, stop, nil
+}
+
+func (m *Metrics) RecordUp() {
```

Comment:
Let's add light docstring comments to each of these functions for more intuitive readability; e.g:
// RecordUp ... Records that the Pessimism application has been successfully stood up